### ISSUE TITLE
setName() is working incorrectly

### ISSUE THREAD

**Issue by videde** [Respect/Validation#213]
# Issue
When I tried test to the code 
```php
$usernameValidator = v::alnum()->noWhitespace()->length(1,15)->setName('username');
try {
    $usernameValidator->assert('really messed up screen#name');
} catch(\InvalidArgumentException $e) {
    var_dump($e->findMessages(array('alnum', 'length', 'noWhitespace')));
}
```
Result is
```
array(3) {
  ["alnum"]=>
  string(79) ""really messed up screen#name" must contain only letters (a-z) and digits (0-9)"
  ["length"]=>
  string(66) ""really messed up screen#name" must have a length between 1 and 15"
  ["noWhitespace"]=>
  string(58) ""really messed up screen#name" must not contain whitespace"
}
```
but I want expected result is when use ```setName('username')``` function
```
array(3) {
  ["alnum"]=>
  string(57) "username must contain only letters (a-z) and digits (0-9)"
  ["length"]=>
  string(44) "username must have a length between 1 and 15"
  ["noWhitespace"]=>
  string(36) "username must not contain whitespace"
}
```
# Reason
I find down why ```setName()``` function is working incorrectly. I use ```var_dump($usernameValidator)``` to reveal the variable ```$usernameValidator``` and there is result
```php
object(Respect\Validation\Validator)#2 (3) {
["rules":protected]=>
  array(3) {
    ["000000002989de9200000000401eb328"]=>
    object(Respect\Validation\Rules\Alnum)#4 (3) {
    ["additionalChars"]=>
      string(0) ""
    ["name":protected]=>
      NULL
      ["template":protected]=>
      NULL
    }
    ["000000002989de9300000000401eb328"]=>
    object(Respect\Validation\Rules\NoWhitespace)#5 (2) {
    ["name":protected]=>
      NULL
      ["template":protected]=>
      NULL
    }
    ["000000002989de9000000000401eb328"]=>
    object(Respect\Validation\Rules\Length)#6 (5) {
    ["minValue"]=>
      int(1)
      ["maxValue"]=>
      int(15)
      ["inclusive"]=>
      bool(true)
      ["name":protected]=>
      NULL
      ["template":protected]=>
      NULL
    }
  }
  ["name":protected]=>
  string(8) "username"
["template":protected]=>
  NULL
}
```
As you see, ```name``` field of ```object(Respect\Validation\Rules\Alnum), object(Respect\Validation\Rules\NoWhitespace), object(Respect\Validation\Rules\Length)``` are ```NULL``` (they should be ```username```), but ```name``` field of ```object(Respect\Validation\Validator)``` is ```username```

# Solution
My solution is that override ```setName()``` function in ```AbstractComposite``` class
```php
abstract class AbstractComposite extends AbstractRule
{
    .
    .
    .

    public function setName($name) {
        foreach ($this->rules as $rule) {
            if ($rule->getName() == NULL)
                $rule->setName($name);
        }
        return parent::setName($name);
    }
}
```

**Comment by Alex-Nabu** [comment]
+1

**Comment by henriquemoody** [comment]
Thanks for reporting it. Pretty good description, BTW.

**Comment by mrsoto** [comment]
Hello,
   I've respect/validation 1.0.0 and this problem persist

**Comment by henriquemoody** [comment]
@mrsoto, this bug as described is fixed since version 0.7.0, you're probably having a similar problem but not this one.

Could you elaborate what you've been experiencing with the library by creating a [new issue](https://github.com/Respect/Validation/issues/new)?
